[metrics] - increment 5

dip96 opened the pull request.

dip96 committed:
[metrics] - increment 1

dip96 committed:
[metrics] - increment 2

dip96 committed:
[metrics] - increment 3

dip96 committed:
[metrics] - increment 4

dip96 committed:
[metrics] - increment 5

dip96 committed:
[metrics] - increment 5

dip96 committed:
[metrics] - increment 5

dip96 committed:
[metrics] - increment 5

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для тестов

dip96 committed:
[metrics] - правки для статик тестов

popoffvg commented on cmd/agent/flags.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// неэкспортированная переменная flagRunAddr содержит адрес и порт для запуска сервера
+var flagRunAddr string
```
Лучше не использовать глобальные переменные, а для конфигурации сделать отедльный тип.

dip96 replied:
Вынес в отдельный тип. Но всё равно использую глобальную переменную - var conf Config, для получения нужных данных. Чтобы избавиться и от неё, у меня пока только одно решение, это возвращать инициализированный в методе parseFlags "conf := Config{}"  объект

popoffvg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,141 @@
+// TODO не до конца понимаю, как реализовать условия из "Важно"
+// TODO переделать ассоциативный масссив на  map[string]interface{}???
```
Если ты имел ввиду про важно в 3-ем инкременте, то приди, пожалуйста, на 1-1, постараюсь уложиться в полчаса. 
Использовать пустой interface{} нужно только в особых случаях. В курсе это может понадобиться только на последних инкрементах.

dip96 replied:
Записался на 19 февраля

popoffvg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,141 @@
+	//не до конца понимаю, как можно связать с http cервером
+	//e := echo.New()
+	//
+	//err := e.Start(flagRunAddr)
```
Найти пример можно тут: https://echo.labstack.com/docs/start-server

dip96 replied:
То есть нужно использовать другой роутер/фреймворк для отправки данных?

popoffvg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,141 @@
+}
+
+func sendMetrics(metrics map[string]map[string]string) {
+	//не понимаю, как отправить запрос используя echo, не поднимая сервер
```
echo - это роутер http запросов со стороны сервера. echo не рассчитан на отправку запроса

dip96 replied:
То есть нужно использовать другой роутер/фреймворк для отправки данных?

popoffvg replied:
Нужно использовать стандартную бибилотеку

popoffvg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,141 @@
+				nil)
+
+			if err != nil {
+				return
```
С ошибкой нужно что-то сделать. Или прекратить обработку и вернуть ошибку, коду, который вызвал sendMetrics. Либо заллогировать. Пропускать не стоит

dip96 replied:
Добавил логирование

popoffvg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,141 @@
+				return
+			}
+
+			errClose := post.Body.Close()
```
тут не стоит ошибку именовать по-другому. 
В go стандартная практика:
```go
err := do1
if err != nil {return err}

err = do2 // обарти внимание на отсутствие :
if err != nil {return err}

err = do3
if err != nil {return err}
```

dip96 replied:
Поправил

popoffvg commented on cmd/server/main.go:
```diff
@@ -1,3 +1,175 @@
+}
+
+// Интерфейс для работы с хранилищем
+type Storage interface {
```
Я забегу вперед и подскажу достаточно расширяемый дизайн для этого задания:
```go
type MetricType string
const(
    MetricTypeGauge MetricType = "gauge"
    MetricTypeCounter MetricType = "counter"
)

type Metric struct{
     Type MetricType
      CounterValue *int
      GaugeValue *float64
}

type Storage interface {
       Get(context.Context, MetricType) (Metric, error)
 	Set(context.Context, Metric) error
 	GetAll(context.Context) ([]metric, error)
 }
```

dip96 replied:
Подскажите, а что должен хранить в себе context?

popoffvg replied:
Нужно дополнительно посмотреть:
type MetricType string
type MetricType = string

popoffvg commented on cmd/server/main.go:
```diff
@@ -1,3 +1,175 @@
+}
+
+// Хранилище метрик
+var storage Storage
```
Не стоит использовать глобальные переменные

dip96 replied:
Подскажите, а как правильно действовать в таком случаи? Я к тому, что данная переменная используется во многих методах, и я пока не понимаю, как её получать не используя глобальную переменную

popoffvg commented on cmd/server/main.go:
```diff
@@ -1,3 +1,175 @@
+
+	metric := storage.Get(nameMetric)
+
+	switch m := metric.(type) {
```
Нужно стараться писать максимально типизированный код.
Но очень круто, что не побоялся поработать с рефлексией

dip96 replied:
Страха не было, потому что на тот момент до конца не понимал, что такое рефлексия. Поправил

dip96 committed:
[metrics] - правка после ревью.
https://github.com/dip96/metrics/pull/5#discussion_r1489967269

dip96 committed:
[metrics] - правка после ревью.
https://github.com/dip96/metrics/pull/5#discussion_r1489967269

dip96 committed:
[metrics] - правка после ревью.
https://github.com/dip96/metrics/pull/5#discussion_r1489980545

https://github.com/dip96/metrics/pull/5#discussion_r1489982584

dip96 committed:
[metrics] - правка после ревью

dip96 committed:
[metrics] - правка после ревью

dip96 committed:
[metrics] - правки статик тестов

dip96 committed:
[metrics] -

dip96 committed:
[metrics] - правки

dip96 committed:
[metrics] - правки после ревью

dip96 committed:
[metrics] - правки после ревью
